### PR TITLE
Adding a Garbage-collection callback for a synchronized garbage-collection across ALL processes

### PR DESCRIPTION
**Pull request by gautham-kollu**
# What does this PR do ?

Adding a Garbage-collection callback for a synchronized garbage-collection across ALL processes

**Collection**: [Note which collection this PR will affect]
Potentially affects all of them.

# Changelog 
- Garbage Collection callback for NeMo 2.0

# Usage
```
    trainer = nl.Trainer(
        strategy = strategy,
        devices=8,
		num_nodes=1,
		accelerator='gpu',
		logger=False,
		enable_checkpointing=False,
		use_distributed_sampler=False,
		#max_epochs=100,
		max_steps=200,
		log_every_n_steps=1,
		val_check_interval=10,
		limit_val_batches=10,
		limit_test_batches=200,
		accumulate_grad_batches=1,
		gradient_clip_val=1.0,
		benchmark=False,
		enable_model_summary=False,
		limit_train_batches=200,
		enable_progress_bar=False,
		num_sanity_val_steps=0,


        **callbacks = [GarbageCollectionCallback(gc_interval_train=20, gc_interval_val=40)],**


        plugins=nl.MegatronMixedPrecision(precision='bf16-mixed',)
    )
```

# Testing 

Log excerpt 
```
      self.communication_handle = torch.distributed._reduce_scatter_base(
    
[NeMo I 2024-08-26 14:07:29 garbage_collection:54] Running garbage collection at train global_step: 20


[NeMo I 2024-08-26 14:07:33 garbage_collection:54] Running garbage collection at train global_step: 40
[NeMo I 2024-08-26 14:07:34 garbage_collection:67] Running garbage collection at validation step: 40
[NeMo I 2024-08-26 14:07:37 garbage_collection:54] Running garbage collection at train global_step: 60
[NeMo I 2024-08-26 14:07:41 garbage_collection:54] Running garbage collection at train global_step: 80
[NeMo I 2024-08-26 14:07:42 garbage_collection:67] Running garbage collection at validation step: 80
[NeMo I 2024-08-26 14:07:45 garbage_collection:54] Running garbage collection at train global_step: 100
[NeMo I 2024-08-26 14:07:48 garbage_collection:54] Running garbage collection at train global_step: 120
[NeMo I 2024-08-26 14:07:49 garbage_collection:67] Running garbage collection at validation step: 120
[NeMo I 2024-08-26 14:07:52 garbage_collection:54] Running garbage collection at train global_step: 140
[NeMo I 2024-08-26 14:07:56 garbage_collection:54] Running garbage collection at train global_step: 160
[NeMo I 2024-08-26 14:07:56 garbage_collection:67] Running garbage collection at validation step: 160
[NeMo I 2024-08-26 14:07:59 garbage_collection:54] Running garbage collection at train global_step: 180
[NeMo I 2024-08-26 14:08:03 garbage_collection:54] Running garbage collection at train global_step: 200
[NeMo I 2024-08-26 14:08:04 garbage_collection:67] Running garbage collection at validation step: 200
`Trainer.fit` stopped: `max_steps=200` reached.
ENDING TIMING RUN AT 2024-08-26 02:08:12 PM
RESULT,large language model,1,79,root,2024-08-26 02:06:53 PM
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/callbacks/garbage_collection.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import gc
+from typing import Any
+
+import pytorch_lightning as pl
+from nemo.utils import logging
+
+
+class GarbageCollectionCallback(pl.Callback):
+    """Callback for synchronized manual Garbage Collection. This is required for distributed training
+    as all processes need to synchronize to garbage collect at the same time, without which
```
Comment (erhoo82) [inline nemo/lightning/pytorch/callbacks/garbage_collection.py:10]: `as all processes`
-> `as all processes on different ranks`

File: nemo/lightning/pytorch/callbacks/garbage_collection.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    one process might hog or straggle all the rest of the processes.
+
+    Migration from NeMo 1.0:
+        When mitrating from NeMo1,
```
Comment (erhoo82) [inline nemo/lightning/pytorch/callbacks/garbage_collection.py:14]: `mitrating` -> `training`?

File: nemo/lightning/pytorch/callbacks/garbage_collection.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            - gc_interval = 0 implied no GC, simply do not add this callback to the trainer
+            - gc_interval > 0, this config is maps => gc_interval_train
+
+            - env-var:NEMO_MANUAL_GC_IN_VALIDATION=0 or doesn't exist => Set gc_interval_val to a very high value that it does not practically run.
+            - env-var:NEMO_MANUAL_GC_IN_VALIDATION=1 => Set gc_interval_val to the same value as gc_interval
+
+        Moving from boolean flag (NEMO_MANUAL_GC_IN_VALIDATION) to integer is to allow user to set any value based on the size of the
```
Comment (erhoo82) [inline nemo/lightning/pytorch/callbacks/garbage_collection.py:21]: `user to set any value`
-> `user to set a specific interval`

File: nemo/lightning/pytorch/callbacks/garbage_collection.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    Note: This callback does not run gc at the start or the end of training or validation.
+    """
+
+    def __init__(self, gc_interval_train, gc_interval_val) -> None:
+        """_summary_
+
+        Args:
+            gc_interval (int, mandatory): Number of global train steps at which garbage collection is done.
+            gc_interval_val (int, mandatory): Number of global validatino steps at which garbage collection is done.
```
Comment (erhoo82) [inline nemo/lightning/pytorch/callbacks/garbage_collection.py:32]: `validatino`
-> `validation`

File: nemo/lightning/pytorch/callbacks/garbage_collection.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        """
+        assert gc_interval_train > 0, "gc_interval_train should be an integer value larger than 0."
+        assert gc_interval_val > 0, "gc_interval_val should be an integer value larger than 0."
+
+        super().__init__()
+        self.gc_interval_train = gc_interval_train
+        self.gc_interval_val = gc_interval_val
+        # As garbage collection is manually controlled, disable automatic garbage collector.
+        gc.disable()
+        # This counter is required as pl does not have a native way of doing this.
```
Comment (erhoo82) [inline nemo/lightning/pytorch/callbacks/garbage_collection.py:42]: `native way of doing this`
-> `native way to track the validation step counter`

**Comment by erhoo82** [comment]
How do we enable the collection callback?

File: nemo/lightning/pytorch/callbacks/garbage_collection.py (hunk 1 above)
Comment (gautham-kollu) [reply nemo/lightning/pytorch/callbacks/garbage_collection.py:10]: Next revison.

File: nemo/lightning/pytorch/callbacks/garbage_collection.py (hunk 2 above)
Comment (gautham-kollu) [reply nemo/lightning/pytorch/callbacks/garbage_collection.py:14]: Migration/Conversion if a NeMo 1.0 user is trying to use this. 

-> "Migrating a Training from NeMo 1.0". Does this explain it better ?

File: nemo/lightning/pytorch/callbacks/garbage_collection.py (hunk 3 above)
Comment (gautham-kollu) [reply nemo/lightning/pytorch/callbacks/garbage_collection.py:21]: Good suggestion. 
Next rev.

File: nemo/lightning/pytorch/callbacks/garbage_collection.py (hunk 5 above)
Comment (gautham-kollu) [reply nemo/lightning/pytorch/callbacks/garbage_collection.py:42]: Good suggestion. Next rev.

File: nemo/lightning/pytorch/callbacks/garbage_collection.py (hunk 4 above)
Comment (gautham-kollu) [reply nemo/lightning/pytorch/callbacks/garbage_collection.py:32]: Next rev.

**Comment by gautham-kollu** [comment]
> How do we enable the collection callback?

Refer to the usage in the PR description.

**Review by erhoo82 — APPROVED** [review]
LGTM
